### PR TITLE
Make federation endpoints more tolerant of trailing slashes for some endpoints

### LINKED ISSUE #3622: Trailing slashes required on some federation endpoints
* [ ] `PUT /send/$txnId`
* [x] ~~`GET /pull`~~ Unused endpoint
* [ ] `GET /event/$eventId`
* [ ] `GET /state/$context`
* [ ] `GET /state_ids/$roomId`
* [ ] `GET /backfill/$context`
* [ ] `GET /groups/$groupId/roles` 
* [ ] `GET /groups/$groupdId/categories`

### PR DESCRIPTION
**Pull request by anoadragon453**
Receiving side of a solution towards #3622.

Client portion here: https://github.com/matrix-org/synapse/pull/4840

Makes Synapse accept zero or more trailing slashes on the federation endpoints mentioned in #3622.

### CODE REVIEW

File: synapse/federation/transport/server.py [hunk 1]
```diff
@@ -312,7 +312,7 @@ def register(self, server):
 
 
 class FederationSendServlet(BaseFederationServlet):
-    PATH = "/send/(?P<transaction_id>[^/]*)/"
+    PATH = "/send/(?P<transaction_id>[^/]*)/*"
```
Comment (richvdh) [inline synapse/federation/transport/server.py]: you want `?` for "0 or 1".

File: synapse/federation/transport/server.py [hunk 2]
```diff
@@ -1100,7 +1100,7 @@ class FederationGroupsCategoryServlet(BaseFederationServlet):
     """Add/remove/get a category in a group
     """
     PATH = (
-        "/groups/(?P<group_id>[^/]*)/categories/(?P<category_id>[^/]+)$"
+        "/groups/(?P<group_id>[^/]*)/categories/(?P<category_id>[^/]+)/*$"
```
Comment (richvdh) [inline synapse/federation/transport/server.py]: this doesn't look right. surely we should not be accepting trailing slashes where we do not currently have them?

Comment (anoadragon453) [reply synapse/federation/transport/server.py]: What do we want exactly? And wasn't the goal to be permissive about trailing slashes everywhere?

**Comment by anoadragon453** [comment]
...good point.

**Comment by anoadragon453** [comment]
This also may be relevant: https://github.com/matrix-org/synapse/blob/aba5eeabd5c5b6dfb955b160a6a6579fe7a1c741/synapse/rest/client/v1/push_rule.py#L33-L36

File: synapse/federation/transport/server.py [hunk 3]
```diff
@@ -1100,7 +1100,7 @@ class FederationGroupsCategoryServlet(BaseFederationServlet):
     """Add/remove/get a category in a group
     """
     PATH = (
-        "/groups/(?P<group_id>[^/]*)/categories/(?P<category_id>[^/]+)"
+        "/groups/(?P<group_id>[^/]*)/categories/(?P<category_id>[^/]+)/?"
```
Comment (richvdh) [inline synapse/federation/transport/server.py]: I'm pretty sure we decided adding an optional / here was incorrect

File: synapse/federation/transport/server.py [hunk 4]
```diff
@@ -1170,7 +1170,7 @@ class FederationGroupsRoleServlet(BaseFederationServlet):
     """Add/remove/get a role in a group
     """
     PATH = (
-        "/groups/(?P<group_id>[^/]*)/roles/(?P<role_id>[^/]+)"
+        "/groups/(?P<group_id>[^/]*)/roles/(?P<role_id>[^/]+)/?"
```
Comment (richvdh) [inline synapse/federation/transport/server.py]: ditto

File: synapse/federation/transport/server.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/federation/transport/server.py]: Correct, we decided to leave out groups for now.

**Pull request merged**